airbyte-cdk/python/airbyte_cdk/sources/streams/call_rate.py

```diff
@@ -0,0 +1,445 @@
+        """
+        self.item = item
+        self.weight = weight
+        self.rate = rate
+        self.time_to_wait = time_to_wait
+        super().__init__(error)
+
+
+class AbstractCallRatePolicy(abc.ABC):
+    """Call rate policy interface.
+    Should be configurable with different rules, like N per M for endpoint X. Endpoint X is matched with APIBudget.
+    """
+
+    @abc.abstractmethod
+    def try_acquire(self, request: Any, weight: int) -> None:
+        """Try to acquire request
+
+        :param request: a request object representing a single call to API
+        :param weight: number of requests to deduct from credit
+        :return:
+        """
+
+    @abc.abstractmethod
+    def update(self, available_calls: Optional[int], call_reset_ts: Optional[datetime.datetime]) -> None:
+        """Update call rate counting with current values
+
+        :param available_calls:
+        :param call_reset_ts:
+        """
+
+
+class FixedWindowCallRatePolicy(AbstractCallRatePolicy):
+    def __init__(self, window: TimeWindow, call_limit: int):
+        self._current_window = TimeWindow(window.start, window.end)
+        self._offset = window.end - window.start
+        self._call_limit = call_limit
+        self._calls_num = 0
+
+    @property
+    def current_window(self) -> TimeWindow:
+        return self._current_window
+
+    def try_acquire(self, request: Any, weight: int) -> None:
+        self._update_current_window()
+
+        if self._calls_num >= self._call_limit:
+            reset_in = self.current_window.end - datetime.datetime.now()
+            error_message = (
+                f"reached maximum number of allowed calls {self._call_limit} " f"per {self._offset} interval, next reset in {reset_in}."
+            )
+            raise CallRateLimitHit(
+                error=error_message,
+                item=request,
+                weight=weight,
+                rate=f"{self._call_limit} per {self._offset}",
+                time_to_wait=reset_in,
+            )
+
+        self._calls_num += 1
+
+    def update(self, available_calls: Optional[int], call_reset_ts: Optional[datetime.datetime]) -> None:
+        """Update call rate counters, by default only reacts to decreasing updates of available_calls and changes to call_reset_ts
+
+        :param available_calls:
+        :param call_reset_ts:
+        """
+        self._update_current_window()
+        current_available_calls = self._call_limit - self._calls_num
+
+        if available_calls is not None and current_available_calls > available_calls:
+            logger.info("got update from api, adjusting available calls from %s to %s", current_available_calls, available_calls)
+            self._calls_num = self._call_limit - available_calls
+
+        if call_reset_ts is not None and call_reset_ts != self.current_window.end:
+            logger.info("got update from api, adjusting reset time from %s to %s", self.current_window.end, call_reset_ts)
```

Comment:
nit: I tend to find f-strings easier to read

Reply:
Me too, but for frequently called code logger formatting is much faster and has minimal cost when logger or level is disabled.